### PR TITLE
Moving proofs about `Injection`, `Surjection` etc. out of `Pointwise`

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
There's been a comment in the `Pointwise` module for `Product` for a while that says that these proofs are misplaced, and I've been meaning to do something about it. As a side note I'm not sure if I'm the only one but these proofs are pretty terrifying, and every time I have looked in the `Pointwise` modules I'm a little intimidated so this'll be good for my peace of mind!

This PR cleans things up by extracting them to `Data.(Product/Sum).Function`. See CHANGELOG for details.  I'm not particularly wedded to the name `Data.(Product/Sum).Function` but it was the best I could come up with. If someone has a better suggestion I'm all ears.

### CODE REVIEW

File: src/Data/Product/Function/.##Setoid.agda# [hunk 1]
```diff
@@ -0,0 +1 @@
+mld46@alde.cl.cam.ac.uk.20111:1550224696
```
Comment (JacquesCarette) [inline src/Data/Product/Function/.##Setoid.agda#]: Did you mean to check this file in?

**Comment by JacquesCarette** [comment]
There are ways to make some of these proofs less terrifying. Some of these proofs already use the "combinator style" that helps.  Some of the code I've written with Amr Sabry for our ESOP 2016 paper does this.  I wouldn't necessarily adopt our syntax, but overall doing just a few core proofs by hand, and then the rest building on those as combinators, seemed to really work.

File: src/Data/Product/Function/.##Setoid.agda# (hunk 1 above)
Comment (MatthewDaggitt) [reply src/Data/Product/Function/.##Setoid.agda#]: Nope!

File: src/Data/Product/Function/Dependent/Propositional/WithK.agda [hunk 2]
```diff
@@ -0,0 +1,38 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Dependent product combinators for propositional equality
+-- preserving functions
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
```
Comment (gallais) [inline src/Data/Product/Function/Dependent/Propositional/WithK.agda]: I believe this shoule be `--with-K`.

**Pull request merged**